exercises/practice/alphametics/spec/alphametics_spec.cr

```diff
@@ -0,0 +1,68 @@
+require "spec"
+require "../src/*"
+
+describe "Alphametics" do
+  it "puzzle with three letters" do
+    puzzle = "I + BB == ILL"
+    solution = {'I' => 1, 'B' => 9, 'L' => 0}
+    Alphametics.solve(puzzle).should eq(solution)
+  end
+
+  pending "solution must have unique value for each letter" do
+    puzzle = "A == B"
+    solution = Hash(String, Int32).new
+    Alphametics.solve(puzzle).should eq(solution)
+  end
+
+  pending "leading zero solution is invalid" do
+    puzzle = "ACA + DD == BD"
+    solution = Hash(String, Int32).new
+    Alphametics.solve(puzzle).should eq(solution)
+  end
+
+  pending "puzzle with two digits final carry" do
+    puzzle = "A + A + A + A + A + A + A + A + A + A + A + B == BCC"
+    solution = {'A' => 9, 'B' => 1, 'C' => 0}
+    Alphametics.solve(puzzle).should eq(solution)
+  end
+
+  pending "puzzle with four letters" do
+    puzzle = "AS + A == MOM"
+    solution = {'A' => 9, 'S' => 2, 'M' => 1, 'O' => 0}
+    Alphametics.solve(puzzle).should eq(solution)
+  end
+
+  pending "puzzle with six letters" do
+    puzzle = "NO + NO + TOO == LATE"
+    solution = {'N' => 7, 'O' => 4, 'T' => 9, 'L' => 1, 'A' => 0, 'E' => 2}
+    Alphametics.solve(puzzle).should eq(solution)
+  end
+
+  pending "puzzle with seven letters" do
+    puzzle = "HE + SEES + THE == LIGHT"
+    solution = {'E' => 4, 'G' => 2, 'H' => 5, 'I' => 0, 'L' => 1, 'S' => 9, 'T' => 7}
+    Alphametics.solve(puzzle).should eq(solution)
+  end
+
+  pending "puzzle with eight letters" do
+    puzzle = "SEND + MORE == MONEY"
+    solution = {'S' => 9, 'E' => 5, 'N' => 6, 'D' => 7, 'M' => 1, 'O' => 0, 'R' => 8, 'Y' => 2}
+    Alphametics.solve(puzzle).should eq(solution)
+  end
+
+  pending "puzzle with ten letters" do
+    puzzle = "AND + A + STRONG + OFFENSE + AS + A + GOOD == DEFENSE"
+    solution = {
+      'A' => 5, 'D' => 3, 'E' => 4, 'F' => 7, 'G' => 8, 'N' => 0, 'O' => 2, 'R' => 1, 'S' => 6, 'T' => 9,
+    }
+    Alphametics.solve(puzzle).should eq(solution)
+  end
+
+  pending "puzzle with ten letters and 199 addends" do
```

Comment:
This test takes ~12 seconds on my machine which I think will time out in the online editor. I ported my solution from Ruby where it takes ~5-6 seconds (which passes the editor). I'm going to see if there are any optimization opportunities in my solution or consider removing this test from the test suite.

Reply:
I can start up the test runner in docker and it took 20 seconds to run with a codespace session with 4 cores. I think that test has to be shortened since at least at the moment there is not a possibility for parallel execution.

I have had similar student encounters where they use combinations, permutation etc and complain about the slow speed.
What makes Crystal slow here is permutation, Ruby can do a huge permutation in a few milliseconds, and Crystal can require seconds, I can dig into why this is happening but this will likely not be fixed for a while.

Reply:
Thanks for the info. I suspected it was something to do with permutations, but I didn't do any profiling to prove it. I'll disable this test for now.